staging/src/k8s.io/sample-apiserver/README.md

```diff
@@ -24,6 +24,59 @@ HEAD of this repo will match HEAD of k8s.io/apiserver, k8s.io/apimachinery, and
 `sample-apiserver` is synced from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/sample-apiserver.
 Code changes are made in that location, merged into `k8s.io/kubernetes` and later synced here.
 
+## Normal Build and Deploy
+
+### Changes to the Types
+
+If you change the API object type definitions in any of the
+`pkg/apis/.../types.go` files then you will need to update the files
+generated from the type definitions.  There is no convenient script to
+do this in the context of `$GOPATH/src/k8s.io/sample-apiserver`.
```

Comment:
doesn't hack/update-codegen.sh work? It should. If not, it's broken.

Reply:
It does not work.  It relies on stuff outside the staging area for sample-apiserver.  This is a non-trivial matter of dependencies.

Reply:
I tried this and it seems to work:
```shell
$ go get k8s.io/sample-apiserver
$ cd $GOPATH/src/k8s.io/sample-apiserver
$ hack/update-codegen.sh
```

Reply:
And I smoke-tested by removing a deepcopy file. After running the script, the file is back. So it also seems to work.

Reply:
Possible it does not work out of the box because we switch to go.mod partially and do not publish a vendor/ dir anymore.

Reply:
Double checked. With godeps it works, with go.mod we miss the dependencies. Probably some `GO111MODULE=on go get k8s.io/...` is necessary.

Reply:
```
$ GO111MODULE=on go mod vendor
# prefix *.sh calls in hack/update-codegen.sh with `bash `
$ hack/update-codegen.sh
```


`go mod vendor` seems to drop executible flags from files. So the bash prefix is necessary. Simple PR. Then 🎉

Reply:
@sttts: when I try starting from scratch, in legacy (not modules) mode, I have gross problems.

```
mjs11:~ mspreitzus.ibm.com$ cd $GOPATH/src
mjs11:src mspreitzus.ibm.com$ mv k8s.io k8s.io-hide
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ printenv | grep GO
GOPATH=/Users/mspreitzus.ibm.com/Documents/go2
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ printenv | grep CODE
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ go get k8s.io/sample-apiserver
# k8s.io/kube-openapi/pkg/schemaconv
k8s.io/kube-openapi/pkg/schemaconv/smd.go:149:55: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:150:20: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:197:52: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:198:11: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:205:11: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:213:11: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:220:12: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:231:12: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:233:40: undefined: schema.UnionField
k8s.io/kube-openapi/pkg/schemaconv/smd.go:240:12: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:240:12: too many errors
# k8s.io/apiserver/pkg/storage/storagebackend/factory
k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:89:3: unknown field 'CAFile' in struct literal of type "github.com/coreos/etcd/pkg/transport".TLSInfo
k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:104:3: cannot use []"google.golang.org/grpc".DialOption literal (type []"google.golang.org/grpc".DialOption) as type []"github.com/coreos/etcd/vendor/google.golang.org/grpc".DialOption in field value
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ go version
go version go1.12.5 darwin/amd64
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ cd k8s.io/sample-apiserver/
mjs11:sample-apiserver mspreitzus.ibm.com$ 
mjs11:sample-apiserver mspreitzus.ibm.com$ hack/update-codegen.sh 
hack/update-codegen.sh: line 28: ../code-generator/generate-groups.sh: No such file or directory
mjs11:sample-apiserver mspreitzus.ibm.com$ 
```

If I then try it in modules mode, I get the same problems.

```
mjs11:sample-apiserver mspreitzus.ibm.com$ cd $GOPATH/src
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ rm -rf k8s.io
mjs11:src mspreitzus.ibm.com$ export GO111MODULES=on
mjs11:src mspreitzus.ibm.com$ printenv | grep GO
GO111MODULES=on
GOPATH=/Users/mspreitzus.ibm.com/Documents/go2
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ go get k8s.io/sample-apiserver
# k8s.io/kube-openapi/pkg/schemaconv
k8s.io/kube-openapi/pkg/schemaconv/smd.go:149:55: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:150:20: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:197:52: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:198:11: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:205:11: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:213:11: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:220:12: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:231:12: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:233:40: undefined: schema.UnionField
k8s.io/kube-openapi/pkg/schemaconv/smd.go:240:12: undefined: schema.Union
k8s.io/kube-openapi/pkg/schemaconv/smd.go:240:12: too many errors
# k8s.io/apiserver/pkg/storage/storagebackend/factory
k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:89:3: unknown field 'CAFile' in struct literal of type "github.com/coreos/etcd/pkg/transport".TLSInfo
k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:104:3: cannot use []"google.golang.org/grpc".DialOption literal (type []"google.golang.org/grpc".DialOption) as type []"github.com/coreos/etcd/vendor/google.golang.org/grpc".DialOption in field value
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ cd k8s.io/sample-apiserver/
mjs11:sample-apiserver mspreitzus.ibm.com$ 
mjs11:sample-apiserver mspreitzus.ibm.com$ hack/update-codegen.sh 
hack/update-codegen.sh: line 28: ../code-generator/generate-groups.sh: No such file or directory
mjs11:sample-apiserver mspreitzus.ibm.com$ 
```

Reply:
I do not know how to do a modules-mode fetch of a whole module, if there is such a thing; `GO111MODULE=on go get $path` is designed to add an import to an existing module, not fetch a module for development on its own.

Based on your comments, I found the following dodge works:
```
mjs11:~ mspreitzus.ibm.com$ cd $GOPATH/src
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ go version
go version go1.12.5 darwin/amd64
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ printenv | grep GO
GOPATH=/Users/mspreitzus.ibm.com/Documents/go2
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ ls
bitbucket.org		github.ibm.com		gopkg.in
fields-left.txt		go.etcd.io		k8s.io-hide
fields-right.txt	golang.org		sigs.k8s.io
github.com		google.golang.org
mjs11:src mspreitzus.ibm.com$ ls sigs.k8s.io/
kind	yaml
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ go get -d k8s.io/sample-apiserver
mjs11:src mspreitzus.ibm.com$ ls
bitbucket.org		github.ibm.com		gopkg.in
fields-left.txt		go.etcd.io		k8s.io
fields-right.txt	golang.org		k8s.io-hide
github.com		google.golang.org	sigs.k8s.io
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ ls k8s.io
api			client-go		kube-openapi
apimachinery		component-base		sample-apiserver
apiserver		klog			utils
mjs11:src mspreitzus.ibm.com$ ls sigs.k8s.io/
kind			structured-merge-diff	yaml
mjs11:src mspreitzus.ibm.com$ 
mjs11:src mspreitzus.ibm.com$ pushd k8s.io/sample-apiserver/
~/Documents/go2/src/k8s.io/sample-apiserver ~/Documents/go2/src
mjs11:sample-apiserver mspreitzus.ibm.com$ 
mjs11:sample-apiserver mspreitzus.ibm.com$ export GO111MODULE=on
mjs11:sample-apiserver mspreitzus.ibm.com$ 
mjs11:sample-apiserver mspreitzus.ibm.com$ go mod vendor
go: finding golang.org/x/tools v0.0.0-20190313210603-aa82965741a9
...
go: extracting gonum.org/v1/gonum v0.0.0-20190331200053-3d26580ed485
mjs11:sample-apiserver mspreitzus.ibm.com$ 
mjs11:sample-apiserver mspreitzus.ibm.com$ ls vendor/
github.com		google.golang.org	modules.txt
golang.org		gopkg.in		sigs.k8s.io
gonum.org		k8s.io
mjs11:sample-apiserver mspreitzus.ibm.com$ 
mjs11:sample-apiserver mspreitzus.ibm.com$ ls vendor/sigs.k8s.io/
structured-merge-diff	yaml
mjs11:sample-apiserver mspreitzus.ibm.com$ 
mjs11:sample-apiserver mspreitzus.ibm.com$ go build
mjs11:sample-apiserver mspreitzus.ibm.com$ 
```

Reply:
I updated this PR to include that tweak to `hack/update-codegen.sh` so that it works even with `go mod vendor`, and updated the README to explain the two options for generating code in the published (not k/k) context.